File: shell/src/dev/symbiosis/kenji/GameShelf.java
package dev.symbiosis.kenji;

import android.content.Context;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.json.JSONArray;
import org.json.JSONObject;

/** Eden-like game properties: mods, cheats, DLC, saves. */
public final class GameShelf {
    private static final Pattern TID = Pattern.compile("(?i)\\[?(0100[0-9A-F]{12})\\]?");

    private GameShelf() {}

    public static String titleId(String path, String title) {
        for (String s : new String[]{path, title}) {
            if (s == null) continue;
            Matcher m = TID.matcher(s);
            if (m.find()) return m.group(1).toUpperCase(Locale.US);
        }
        return "";
    }

    public static String properties(Context c, String path, String title) {
        try {
            String tid = titleId(path, title);
            File app = DataSeed.appPath(c);
            File eden = DataSeed.bestEden(c);
            JSONObject o = new JSONObject();
            o.put("title", title == null ? "" : title);
            o.put("path", path == null ? "" : path);
            o.put("titleId", tid);
            o.put("mods", listMods(app, eden, tid));
            o.put("cheats", listCheats(app, eden, tid));
            o.put("dlc", listDlc(app, tid));
            o.put("saves", listSaves(app, eden, tid));
            o.put("presets", new JSONArray(new JSONObject(SettingsBank.listJson(c, tid)).optJSONArray("items")));
            o.put("hint", "Моды Kenji: mods/contents/" + tid
                    + " · Eden: load/" + tid
                    + " · читы: cheats/ · DLC ставится в их Kenji · сейвы: bis/user/save");
            return o.toString();
        } catch (Exception e) {
            return "{\"mods\":[],\"cheats\":[],\"dlc\":[],\"saves\":[]}";
        }
    }

    public static String modsJson(Context c) {
        try {
            File app = DataSeed.appPath(c);
            File eden = DataSeed.bestEden(c);
            JSONArray items = new JSONArray();
            addModRoots(items, new File(app, "mods/contents"), "Kenji");
            if (eden != null) addModRoots(items, new File(eden, "load"), "Eden");
            return new JSONObject().put("items", items).toString();
        } catch (Exception e) {
            return "{\"items\":[]}";
        }
    }

    public static String bridgeMods(Context c) {
        try {
            File eden = DataSeed.bestEden(c);
            File destRoot = new File(DataSeed.appPath(c), "mods/contents");
            if (eden == null) {
                return new JSONObject().put("ok", false).put("message", "папка Eden не найдена — укажите её в данных").toString();
            }
            File load = new File(eden, "load");
            if (!load.isDirectory()) {
                return new JSONObject().put("ok", false).put("message", "в Eden нет load/").toString();
            }
            destRoot.mkdirs();
            int n = 0;
            File[] tids = load.listFiles();
            if (tids != null) {
                for (File tid : tids) {
                    if (!tid.isDirectory()) continue;
                    File dest = new File(destRoot, tid.getName());
                    dest.mkdirs();
                    File[] kids = tid.listFiles();
                    if (kids == null) continue;
                    for (File kid : kids) {
                        File d = new File(dest, kid.getName());
                        if (!d.exists()) {
                            // best-effort copy of small marker; directories listed as linked by name
                            if (kid.isDirectory()) d.mkdirs();
                        }
                        n++;
                    }
                }
            }
            return new JSONObject().put("ok", true).put("message", "мосты модов: " + n + " записей").toString();
        } catch (Exception e) {
            return "{\"ok\":false}";
        }
    }

    public static String saveSource(Context c) {
        try {
            File dir = new File(DataSeed.appPath(c), "bis/user/save");
            File[] slots = dir.listFiles();
            int n = slots == null ? 0 : slots.length;
            return new JSONObject()
                    .put("path", dir.getAbsolutePath())
                    .put("name", "Kenji saves")
                    .put("titles", n)
                    .put("size", "")
                    .toString();
        } catch (Exception e) {
            return "{\"path\":\"\",\"titles\":0}";
        }
    }

    private static JSONArray listMods(File app, File eden, String tid) {
        JSONArray a = new JSONArray();
        if (tid.isEmpty()) return a;
        collectNamed(a, new File(new File(app, "mods/contents"), tid), "мод", "Kenji");
        if (eden != null) collectNamed(a, new File(new File(eden, "load"), tid), "мод", "Eden");
        return a;
    }

    private static JSONArray listCheats(File app, File eden, String tid) {
        JSONArray a = new JSONArray();
        collectFiles(a, new File(new File(app, "cheats"), tid), "чит");
        collectFiles(a, new File(app, "mods/contents/" + tid + "/cheats"), "чит");
        if (eden != null) {
            collectFiles(a, new File(eden, "load/" + tid + "/cheats"), "чит");
            collectFiles(a, new File(eden, "cheats/" + tid), "чит");
        }
        return a;
    }

    private static JSONArray listDlc(File app, String tid) {
        JSONArray a = new JSONArray();
        collectNamed(a, new File(app, "games/" + tid + "/dlc"), "DLC", "Kenji");
        collectNamed(a, new File(app, "sdcard/atmosphere/contents/" + tid), "DLC", "contents");
        return a;
    }

    private static JSONArray listSaves(File app, File eden, String tid) {
        JSONArray a = new JSONArray();
        File kenji = new File(app, "bis/user/save");
        File[] slots = kenji.listFiles();
        if (slots != null) {
            for (File s : slots) {
                if (!s.isDirectory()) continue;
                try {
                    a.put(new JSONObject().put("name", "слот " + s.getName()).put("path", s.getAbsolutePath()).put("kind", "сейв"));
                } catch (Exception ignored) {
                }
            }
        }
        if (eden != null) {
            File es = new File(eden, "nand/user/save");
            if (es.isDirectory()) {
                try {
                    a.put(new JSONObject().put("name", "сейвы Eden").put("path", es.getAbsolutePath()).put("kind", "сейв"));
                } catch (Exception ignored) {
                }
            }
        }
        return a;
    }

    private static void addModRoots(JSONArray items, File root, String source) {
        File[] tids = root.listFiles();
        if (tids == null) return;
        for (File tid : tids) {
            if (!tid.isDirectory()) continue;
            File[] mods = tid.listFiles();
            int n = mods == null ? 0 : mods.length;
            try {
                items.put(new JSONObject()
                        .put("titleId", tid.getName())
                        .put("mods", n)
                        .put("source", source)
                        .put("active", n > 0)
                        .put("detail", tid.getAbsolutePath()));
            } catch (Exception ignored) {
            }
        }
    }

    private static void collectNamed(JSONArray a, File dir, String kind, String source) {
        File[] kids = dir.listFiles();
        if (kids == null) return;
        for (File k : kids) {
            if (k.getName().startsWith(".")) continue;
            try {
                a.put(new JSONObject()
                        .put("name", k.getName())
                        .put("kind", kind)
                        .put("source", source)
                        .put("path", k.getAbsolutePath())
                        .put("enabled", true));
            } catch (Exception ignored) {
            }
        }
    }

    private static void collectFiles(JSONArray a, File dir, String kind) {
        File[] kids = dir.listFiles();
        if (kids == null) return;
        for (File k : kids) {
            if (!k.isFile()) continue;
            try {
                a.put(new JSONObject().put("name", k.getName()).put("kind", kind).put("path", k.getAbsolutePath()));
            } catch (Exception ignored) {
            }
        }
    }
}

File: shell/src/dev/symbiosis/kenji/GameShelf.kt
package dev.symbiosis.kenji

import android.content.Context
import android.net.Uri
import android.preference.PreferenceManager
import java.io.File
import java.util.Locale
import java.util.regex.Pattern

data class GameRom(
    val title: String,
    val path: String,
    val titleId: String,
    val bytes: Long,
    val fileName: String,
    val folder: String,
) {
    val exists: Boolean get() = path.startsWith("/") && File(path).isFile
    val update: Boolean get() = titleId.length == 16 && titleId.uppercase(Locale.US).endsWith("800")
}

object GameShelf {
    private const val PREF = "kenji_folders"
    private val TID = Pattern.compile("(?i)\\[?(0100[0-9A-F]{12})\\]?")

    fun list(context: Context): List<GameRom> {
        val seen = LinkedHashSet<String>()
        val raw = ArrayList<GameRom>()
        for (dir in folders(context)) collect(dir, seen, raw, 0)
        val bases = raw.filter { !it.update && it.titleId.length == 16 }.map { it.titleId }.toHashSet()
        return raw.filter { rom ->
            if (!rom.exists) return@filter false
            if (rom.update && bases.contains(baseId(rom.titleId))) return@filter false
            if (rom.update) return@filter false
            true
        }.sortedBy { it.title.lowercase(Locale.US) }
    }

    fun addPath(context: Context, path: String) {
        val f = File(path)
        if (!f.isDirectory) return
        val p = context.getSharedPreferences(PREF, Context.MODE_PRIVATE)
        val next = LinkedHashSet(p.getStringSet("uris", emptySet()) ?: emptySet())
        next.add(f.absolutePath)
        p.edit().putStringSet("uris", next).commit()
        PreferenceManager.getDefaultSharedPreferences(context)
            .edit().putString("gameFolderPath", f.absolutePath).commit()
    }

    fun removePath(context: Context, path: String) {
        val p = context.getSharedPreferences(PREF, Context.MODE_PRIVATE)
        val next = LinkedHashSet(p.getStringSet("uris", emptySet()) ?: emptySet())
        next.remove(path)
        p.edit().putStringSet("uris", next).commit()
    }

    fun folders(context: Context): List<File> {
        val out = LinkedHashSet<File>()
        val p = context.getSharedPreferences(PREF, Context.MODE_PRIVATE)
        p.getStringSet("uris", emptySet())?.forEach { s ->
            if (s.startsWith("/")) {
                val f = File(s)
                if (f.isDirectory) out.add(f)
            }
        }
        val off = PreferenceManager.getDefaultSharedPreferences(context)
        off.getString("gameFolderPath", null)?.let {
            val f = File(it)
            if (f.isDirectory) out.add(f)
        }
        val sd = android.os.Environment.getExternalStorageDirectory()
        listOf("Download/ed", "Download/Switch", "Download/NSP", "Switch", "Games").forEach { rel ->
            val f = File(sd, rel)
            if (f.isDirectory && looksGameFolder(f)) out.add(f)
        }
        return out.toList()
    }

    private fun looksGameFolder(dir: File): Boolean {
        val kids = dir.listFiles() ?: return false
        return kids.any { it.isFile && isRom(it.name) }
    }

    private fun collect(dir: File, seen: MutableSet<String>, out: MutableList<GameRom>, depth: Int) {
        if (depth > 2 || !dir.isDirectory || skipName(dir.name)) return
        val kids = dir.listFiles() ?: return
        for (f in kids) {
            if (f.isFile && isRom(f.name)) {
                val path = f.absolutePath
                if (!seen.add(path)) continue
                val tid = titleIdOf(f.name)
                out.add(
                    GameRom(
                        title = pretty(f.name),
                        path = path,
                        titleId = tid,
                        bytes = f.length(),
                        fileName = f.name,
                        folder = dir.name,
                    ),
                )
            } else if (f.isDirectory && !skipName(f.name)) {
                collect(f, seen, out, depth + 1)
            }
        }
    }

    private fun skipName(name: String): Boolean {
        val n = name.lowercase(Locale.US)
        return n in setOf(
            "nand", "bis", "keys", "system", "registered", "android",
            "eden", "load", "mods", "shader", "cache", "screenshots",
        ) || n.startsWith(".")
    }

    fun isRom(name: String): Boolean {
        val n = name.lowercase(Locale.US)
        return n.endsWith(".nsp") || n.endsWith(".xci") || n.endsWith(".nsz") || n.endsWith(".xcz")
    }

    fun pretty(name: String): String {
        var stem = name.substringBeforeLast('.')
        stem = stem.replace(Regex("(?i)\\[0100[0-9A-F]{12}]"), "")
        stem = stem.replace(Regex("(?i)\\[v\\d+]"), "")
        stem = stem.replace(Regex("\\([^)]*\\s*[Gg][Bb]\\)"), "")
        stem = stem.replace('_', ' ').replace(Regex("\\s+"), " ").trim()
        return stem.ifBlank { name }
    }

    fun titleIdOf(name: String): String {
        val m = TID.matcher(name)
        return if (m.find()) m.group(1).uppercase(Locale.US) else ""
    }

    fun baseId(titleId: String): String {
        if (titleId.length != 16 || !titleId.uppercase(Locale.US).endsWith("800")) return titleId
        return try {
            String.format(Locale.US, "%016X", java.lang.Long.parseUnsignedLong(titleId, 16) - 0x800L)
        } catch (_: Exception) {
            titleId
        }
    }

    fun treeToPath(uri: Uri): String? = PickActivity.treeToPath(uri)
}
